streamz/sources.py

```diff
@@ -148,6 +148,154 @@ def do_poll(self):
                 break
 
 
+@Stream.register_api(staticmethod)
+class from_socket(Source):
+    """
+    Creates events by reading from a socket.
+
+    The stream of incoming bytes is split on a given delimiter, and the parts
+    become the emitted events.
+
+    Based on example at
+    https://www.tornadoweb.org/en/stable/ioloop.html#tornado.ioloop.IOLoop
+
+    Parameters
+    ----------
+    port : int
+        The port to open and listen on. It only gets opened when the source
+        is started, and closed upon ``stop()``
+    delimiter : bytes
+        The incoming data will be split on this value. The resulting events
+        will still have the delimiter at the end.
+    start : bool
+        Whether to immediately initiate the source. You probably want to
+        set up downstream nodes first.
+
+    Example
+    -------
+
+    >>> source = Source.from_socket(4567)  # doctest: +SKIP
+    """
+    def __init__(self, port, delimiter=b'\n', start=False):
+        self.delimiter = delimiter
+        super(from_socket, self).__init__(ensure_io_loop=True)
+        self.stopped = True
+        self.port = port
+        self.sock = None
+        if start:
+            self.start()
+
+    def start(self):
+        """Sets up socket and puts connection handler on event loop"""
+        import socket
+        if self.stopped:
+            self.stopped = False
+            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
+            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            self.sock.setblocking(0)
+            self.sock.bind(("", self.port))
+            self.sock.listen(128)
+            self.loop.add_handler(self.sock.fileno(), self.connection_ready,
+                                  self.loop.READ)
```

Comment:
I'm somewhat surprised to see the use of raw sockets here.  I'm curious, was there a reason to go this route rather than create a Tornado IOStream here directly?  I see that you use them later on.  

I don't know a ton about managing nonblocking sockets manually, I'm probably just missing something.

Reply:
This code is adapted from the tornado IOLoop documentation: as far as I understand it, you do not have any IOStream until you accept an incoming connection on the open socket.

Reply:
Ah, maybe the right abstraction here then is a Tornado TCPServer?